test/unit/cmap/connection.test.ts

```diff
@@ -0,0 +1,248 @@
+    };
+
+    // @ts-expect-error: This subset of options is all that is needed
+    connect(options, (err, conn) => {
+      expect(err).to.be.a('undefined');
+      expect(conn).to.be.instanceOf(Connection);
+      expect(conn).to.have.property('hello').that.is.a('object');
+
+      conn.command(ns('$admin.cmd'), { ping: 1 }, { socketTimeoutMS: 50 }, err => {
+        const beforeHandshakeSymbol = getSymbolFrom(err, 'beforeHandshake', false);
+        expect(beforeHandshakeSymbol).to.be.a('symbol');
+        expect(err).to.have.property(beforeHandshakeSymbol, false);
+
+        done();
+      });
+    });
+  });
+
+  it('should throw a network error with kBeforeHandshake set to true on timeout before hand shake', function (done) {
+    server.setMessageHandler(() => {
+      // respond to no requests to trigger timeout event
+    });
+
+    const options = {
+      hostAddress: server.hostAddress(),
+      socketTimeoutMS: 50
+    };
+
+    // @ts-expect-error: This subset of options is all that is needed
+    connect(options, (err, conn) => {
+      expect(conn).to.be.a('undefined');
+
+      const beforeHandshakeSymbol = getSymbolFrom(err, 'beforeHandshake');
+      expect(err).to.have.property(beforeHandshakeSymbol, true);
+
+      done();
+    });
+  });
+
+  describe('onTimeout()', () => {
+    let connection: sinon.SinonSpiedInstance<Connection>;
+    let clock: sinon.SinonFakeTimers;
+    let driverSocket: sinon.SinonSpiedInstance<FakeSocket>;
+    let messageStream: MessageStream;
+    let kDelayedTimeoutId: symbol;
+    let NodeJSTimeoutClass: any;
+
+    /** The absolute minimum socket API needed by Connection as of writing this test */
+    class FakeSocket extends EventEmitter {
+      address() {
+        // is never called
+      }
+      pipe() {
+        // does not need to do anything
+      }
+      destroy() {
+        // is called, has no side effects
+      }
+      get remoteAddress() {
+        return 'iLoveJavaScript';
+      }
+      get remotePort() {
+        return 123;
+      }
+    }
+
+    beforeEach(() => {
+      clock = sinon.useFakeTimers();
+
+      NodeJSTimeoutClass = setTimeout(() => null, 1).constructor;
+
+      driverSocket = sinon.spy(new FakeSocket());
+      // @ts-expect-error: This subset of options is all that is needed
+      connection = sinon.spy(new Connection(driverSocket, { id: 1 }));
+      const messageStreamSymbol = getSymbolFrom(connection, 'messageStream');
+      kDelayedTimeoutId = getSymbolFrom(connection, 'delayedTimeoutId');
+      messageStream = connection[messageStreamSymbol];
+    });
+
+    afterEach(() => {
+      clock.restore();
+    });
+
+    it('should delay timeout errors by one tick', async () => {
+      expect(connection).to.have.property(kDelayedTimeoutId, null);
+
+      driverSocket.emit('timeout');
+      expect(connection.onTimeout).to.have.been.calledOnce;
+      expect(connection).to.have.property(kDelayedTimeoutId).that.is.instanceOf(NodeJSTimeoutClass);
+
```

Comment:
can we add an expect connection `closed` to be `false` and expect `driverSocket.destroy` to not have been called to explicitly verify that it doesn't fire sooner?